### PR TITLE
Save/restore column sizes of the tables in the Peers tab

### PR DESCRIPTION
**Pull request by hebasto**

### CODE REVIEW

**Review by jarolrod — COMMENTED** [review]
ACK dda2f65d754aa05d8191e79e4db2189e79988443, tested on macOS 11.2.3 Qt 5.15.2

This PR successfully saves/restores the column sizes in the peers tab. Below are some screenshots showing the behavior:

**Start**
Screen Shot 2021-03-25 at 10 42 18 PM

**Resize**
Screen Shot 2021-03-25 at 10 42 42 PM

**Close and Restart**
Screen Shot 2021-03-25 at 10 43 05 PM

**Review by jonatack — APPROVED** [review]
Very practical improvement.

ACK dda2f65d754aa05d8191e79e4db2189e79988443 code review, debug build is clean, tested repeatedly with both "Peers" and "Banned Peers" column widths and different shutdown methods. Tested with Qt 5.15.2 on Debian 5.10.19-1 (2021-03-02).

Changes are straightforward. I'm not a Qt expert so I briefly checked the Qt 5.15 `QByteArray`, `setState` and `restoreValue` docs and version compats.

**Comment by hebasto** [comment]
@promag 

Mind looking into this PR?

File: src/qt/rpcconsole.cpp [hunk 1]
```diff
@@ -1186,6 +1198,11 @@ void RPCConsole::showEvent(QShowEvent *event)
 
 void RPCConsole::hideEvent(QHideEvent *event)
 {
+    // It is too late to call QHeaderView::saveState() in ~RPCConsole(), as all of
```
Comment (promag) [inline src/qt/rpcconsole.cpp:1201]: An alternative could be to use `QHeaderView::sectionResized` signal.

File: src/qt/rpcconsole.cpp [hunk 2]
```diff
@@ -612,9 +615,14 @@ void RPCConsole::setClientModel(ClientModel *model, int bestblock_height, int64_
         ui->peerWidget->setSelectionBehavior(QAbstractItemView::SelectRows);
         ui->peerWidget->setSelectionMode(QAbstractItemView::ExtendedSelection);
         ui->peerWidget->setContextMenuPolicy(Qt::CustomContextMenu);
-        ui->peerWidget->setColumnWidth(PeerTableModel::Address, ADDRESS_COLUMN_WIDTH);
-        ui->peerWidget->setColumnWidth(PeerTableModel::Subversion, SUBVERSION_COLUMN_WIDTH);
-        ui->peerWidget->setColumnWidth(PeerTableModel::Ping, PING_COLUMN_WIDTH);
+
+        QSettings settings;
+        m_peer_widget_header_state = settings.value("PeersTabPeerHeaderState").toByteArray();
```
Comment (promag) [inline src/qt/rpcconsole.cpp:620]: Load `m_peer_widget_header_state` in constructor where `settings` already exists.

**Review by promag — COMMENTED** [review]
Tested ACK dda2f65d754aa05d8191e79e4db2189e79988443.

File: src/qt/rpcconsole.cpp (hunk 2 above)
Comment (hebasto) [reply src/qt/rpcconsole.cpp:620]: @promag 

Thanks for raising this question!

From my [experience](https://github.com/bitcoin-core/gui/pull/205#issuecomment-807812020) `QTableView::setColumnWidth` does not work without model set (but it is not documented by Qt).

Did I miss something? Does it work for you?

Comment (promag) [reply src/qt/rpcconsole.cpp:620]: I mean move L620 to constructor but leave `restoreState()` here.

File: src/qt/rpcconsole.cpp (hunk 1 above)
Comment (hebasto) [reply src/qt/rpcconsole.cpp:1201]: Yes, using `QHeaderView::sectionResized` signal is more clear solution. But most of them are redundant, as we are interested in final state only.

Do you think we should prefer clear code (I admit that using a comment to explain code is a bad sign)?
If so, I'll be happy to implement your suggestion.

Comment (promag) [reply src/qt/rpcconsole.cpp:1201]: On the other hand with current approach it always save even if you don’t resize.

Is the signal emitted while resizing? Not only at the end? (Didn’t test)

Comment (hebasto) [reply src/qt/rpcconsole.cpp:1201]: @promag
> Is the signal emitted while resizing? Not only at the end? (Didn’t test)

Signal is emitted while resizing, at least on Linux Mint 20.1 + Qt 5.12.8.

**Comment by hebasto** [comment]
Updated dda2f65d754aa05d8191e79e4db2189e79988443 -> 002d0e0d998d4c15d86990f9ce41d981e0bad7fc ([pr256.01](https://github.com/hebasto/gui/commits/pr256.01) -> [pr256.02](https://github.com/hebasto/gui/commits/pr256.02), [diff](https://github.com/hebasto/gui/compare/pr256.01..pr256.02)):

- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/256#discussion_r619304856)

File: src/qt/rpcconsole.cpp (hunk 2 above)
Comment (hebasto) [reply src/qt/rpcconsole.cpp]: Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/256#issuecomment-829545241).

**Review by promag — COMMENTED** [review]
Code review ACK 002d0e0d998d4c15d86990f9ce41d981e0bad7fc.

**Comment by hebasto** [comment]
@jonatack @jarolrod 

Do you mind re-reviewing this PR?

**Comment by hebasto** [comment]
Rebased 002d0e0d998d4c15d86990f9ce41d981e0bad7fc -> 03370446f7fc4936596c7b0d689ab425bdbc825d ([pr256.02](https://github.com/hebasto/gui/commits/pr256.02) -> [pr256.03](https://github.com/hebasto/gui/commits/pr256.03)) due to the conflict with #194.

**Review by jarolrod — COMMENTED** [review]
ACK 03370446f7fc4936596c7b0d689ab425bdbc825d

patch looks good

| Start           | Resize         | Restart         |
| ----------- | ----------- | ------------ |
| Screen Shot 2021-05-12 at 10 41 14 PM | Screen Shot 2021-05-12 at 10 41 40 PM | Screen Shot 2021-05-12 at 10 43 35 PM |

**Comment by hebasto** [comment]
@jonatack @promag 

Hoping on confirming your ACKs after the recent rebasing :)

**Comment by hebasto** [comment]
Rebased 03370446f7fc4936596c7b0d689ab425bdbc825d -> fb1b1e0f3eae32b087347889ed7502b7f2c48549 ([pr256.03](https://github.com/hebasto/gui/commits/pr256.03) -> [pr256.04](https://github.com/hebasto/gui/commits/pr256.04)) due to the conflict with #123.

**Review by jonatack — COMMENTED** [review]
This is a real UX improvement, thanks!

ACK fb1b1e0f3eae32b087347889ed7502b7f2c48549 code review, debug-built and tested

I am seeing some unusual flickering in the `User Agent` header, but this is also the case in current master 38ab7d0765e52b8. I don't recall seeing it before but am not sure.

**Comment by hebasto** [comment]
@jonatack 
> I am seeing some unusual flickering in the `User Agent` header, but this is also the case in current master [38ab7d0](https://github.com/bitcoin-core/gui/commit/38ab7d0765e52b8c9d93a677de70736a2e3a58a0). I don't recall seeing it before but am not sure.

Reported in #191, fixed in #164.

**Review by jarolrod — COMMENTED** [review]
ACK fb1b1e0f3eae32b087347889ed7502b7f2c48549

Code Review ACK and Tested ACK on Linux Qt 5.15.2, macOS 11.3 Qt 5.15.2, and cross-compile to Windows

**Arch Linux:**
| Start      | Resize   | Restart |
| ----------- | ----------- |----------- |
| before-resize | after-resize | restart |

**macOS:**
| Start      | Resize   | Restart |
| ----------- | ----------- |----------- |
| before-resize | after-resize | restart |

**Windows:**
| Start      | Resize   | Restart |
| ----------- | ----------- |----------- |
| before-resize | after-resize | restart |

**Pull request merged**